### PR TITLE
Implement Dropspenser Hook

### PR DESCRIPTION
**Pull request by KrystilizeNevaDies**
Implemented a hook for dropspensing.
HOOK_DROPSPENSE

MyOnDropSpenser(World, DropSpenser, SlotNum)

When plugin callback returns true then dropspense is cancelled

How can i add documentation for this for [api.cuberite.org](https://api.cuberite.org)?

### CODE REVIEW

File: src/Bindings/PluginManager.h [hunk 1]
```diff
@@ -249,7 +251,8 @@ class cPluginManager
 	bool CallHookHandshake                (cClientHandle & a_ClientHandle, const AString & a_Username);
 	bool CallHookHopperPullingItem        (cWorld & a_World, cHopperEntity & a_Hopper, int a_DstSlotNum, cBlockEntityWithItems & a_SrcEntity, int a_SrcSlotNum);
 	bool CallHookHopperPushingItem        (cWorld & a_World, cHopperEntity & a_Hopper, int a_SrcSlotNum, cBlockEntityWithItems & a_DstEntity, int a_DstSlotNum);
-	bool CallHookKilled		      (cEntity & a_Victim, TakeDamageInfo & a_TDI, AString & a_DeathMessage);
+	bool CallHookDropSpense               (cWorld & a_World, cDropSpenserEntity & a_DropSpenser, int a_SlotNum);
+	bool CallHookKilled		      		  (cEntity & a_Victim, TakeDamageInfo & a_TDI, AString & a_DeathMessage);
```
Comment (peterbell10) [inline src/Bindings/PluginManager.h:255]: Spaces for alignment:
```suggestion
	bool CallHookKilled                   (cEntity & a_Victim, TakeDamageInfo & a_TDI, AString & a_DeathMessage);
```

**Review by peterbell10 — CHANGES_REQUESTED** [review]
>How can i add documentation for this for api.cuberite.org?

Documentation goes in the APIDump plugin. For hooks, that's in `Server/Plugins/APIDump/Hooks/`.

**Comment by KrystilizeNevaDies** [comment]
Updated with API documentation

File: src/BlockEntities/DropSpenserEntity.cpp [hunk 2]
```diff
@@ -80,8 +80,16 @@ void cDropSpenserEntity::DropSpense(cChunk & a_Chunk)
 
 	int RandomSlot = 	m_World->GetTickRandomNumber(SlotsCnt - 1);
 
+	int SpenseSlot = OccupiedSlots[RandomSlot];
+
+	if (cPluginManager::Get()->CallHookDropSpense(*m_World, *this, SpenseSlot))
+		{
+			// Plugin disagrees with the move
+			return;
+		}
```
Comment (bearbin) [inline src/BlockEntities/DropSpenserEntity.cpp:89]: Indentation is off here as well.

**Review by bearbin — APPROVED** [review]

File: src/BlockEntities/DropSpenserEntity.cpp [hunk 3]
```diff
@@ -80,8 +80,16 @@ void cDropSpenserEntity::DropSpense(cChunk & a_Chunk)
 
 	int RandomSlot = 	m_World->GetTickRandomNumber(SlotsCnt - 1);
 
+	int SpenseSlot = OccupiedSlots[RandomSlot];
+
+	if (cPluginManager::Get()->CallHookDropSpense(*m_World, *this, SpenseSlot))
```
Comment (bearbin) [inline src/BlockEntities/DropSpenserEntity.cpp:85]: clang-tidy failed here, you need to include PluginManager in this file.

**Review by bearbin — CHANGES_REQUESTED** [review]

**Comment by KrystilizeNevaDies** [comment]
How can i fix this travis-ci error?

`ERROR: Newly undocumented API symbols found:
cPluginManager.HOOK_DROPSPENSE (constant)
The command "./travisbuild.sh" exited with 1.
cache.2
store build cache`

File: Server/Plugins/APIDump/Classes/Plugins.lua [hunk 4]
```diff
@@ -790,6 +790,10 @@ cPluginManager.AddHook(cPluginManager.HOOK_CHAT, OnChatMessage);
 			{
 				Notes = "Called when a hopper is pushing an item into the container it is aimed at.",
 			},
+            HOOK_DROPSPENSE =
```
Comment (bearbin) [inline Server/Plugins/APIDump/Classes/Plugins.lua:793]: Indentation is off here again.

**Comment by bearbin** [comment]
> How can i fix this travis-ci error?
> 
> `ERROR: Newly undocumented API symbols found: cPluginManager.HOOK_DROPSPENSE (constant) The command "./travisbuild.sh" exited with 1. cache.2 store build cache`

You just did, I told you how in the previous review :)

**Comment by KrystilizeNevaDies** [comment]
> > How can i fix this travis-ci error?
> > `ERROR: Newly undocumented API symbols found: cPluginManager.HOOK_DROPSPENSE (constant) The command "./travisbuild.sh" exited with 1. cache.2 store build cache`
> 
> You just did, I told you how in the previous review :)

:)

**Review by bearbin — APPROVED** [review]

**Pull request merged**